java-checks/src/main/java/org/sonar/java/checks/tests/ParameterizedTestCheck.java

```diff
@@ -0,0 +1,160 @@
+      return;
+    }
+
+    Set<MethodTree> handled = new HashSet<>();
+    for (int i = 0; i < methods.size(); i++) {
+      MethodTree method = methods.get(i);
+      List<StatementTree> methodBody = method.block().body();
+      // In addition to filtering literals, we want to count the number of differences since they will represent the number of parameter
+      // that would be required to transform the tests to a single parametrized one.
+      CollectAndFilter collectAndFilter = new CollectAndFilter();
+
+      List<MethodTree> equivalentMethods = methods.stream()
+        .skip(i + 1L)
+        // avoid reporting multiple times
+        .filter(otherMethod -> !handled.contains(otherMethod))
+        // only consider method syntactically equivalent, ignoring literals
+        .filter(otherMethod -> {
+          boolean areEquivalent = SyntacticEquivalence.areEquivalent(methodBody, otherMethod.block().body(), collectAndFilter);
+          if (areEquivalent) {
+            // If methods where not equivalent, we don't want to pollute the set of node to parameterize.
+            collectAndFilter.validateCollection();
+          }
+          return areEquivalent;
+        })
+        .collect(Collectors.toList());
+
+      if (equivalentMethods.size() + 1 >= MIN_SIMILAR_METHODS) {
+        handled.add(method);
+        handled.addAll(equivalentMethods);
+
+        if (collectAndFilter.nodeToParametrize.size() > MAX_NUMBER_PARAMETER) {
+          // We don't report an issue if the change would result in too many parameters.
+          // We still add it to "handled" to not report a subset of candidate methods.
+          continue;
+        }
+
+        List<JavaFileScannerContext.Location> secondaries = collectAndFilter.nodeToParametrize.stream().map(param ->
+          new JavaFileScannerContext.Location("Value to parameterize", param)).collect(Collectors.toCollection(ArrayList::new));
+
+        equivalentMethods.stream().map(equivalentMethod ->
+          new JavaFileScannerContext.Location("Related test", equivalentMethod.simpleName()))
+          .forEach(secondaries::add);
+
+        reportIssue(method.simpleName(), String.format(MESSAGE, equivalentMethods.size() + 1) , secondaries, null);
+      }
+    }
+  }
+
+  private static boolean isParametrizedCandidate(MethodTree methodTree) {
+    BlockTree block = methodTree.block();
+    SymbolMetadata symbolMetadata = methodTree.symbol().metadata();
+    return block != null &&
+      block.body().size() >= MIN_NUMBER_STATEMENTS &&
+      TEST_ANNOTATIONS.stream().anyMatch(symbolMetadata::isAnnotatedWith);
+  }
+
+  static class CollectAndFilter implements BiPredicate<JavaTree, JavaTree> {
+
+    Set<JavaTree> nodeToParametrize = new HashSet<>();
+    private final Set<JavaTree> currentNodeToParameterize = new HashSet<>();
+
+    @Override
+    public boolean test(JavaTree leftNode, JavaTree rightNode) {
+      if (isLiteral(leftNode) && isLiteral(rightNode)) {
```

Comment:
I disagree, the logic should be careful with the semantic kind of the literals, let's take this example:
```
class FooTest {
  @Test
  void test1() { // FP "compute(1)" and "compute(2.0f)" is not related to the same method
    int a = compute(1);
    assertEquals(a, 0);
  }

  @Test
  void test2() {
    int a = compute(2.0f);
    assertEquals(a, 0);
  }

  @Test
  void test3() {
    int a = compute(3);
    assertEquals(a, 0);
  }

  int compute(int a) { return 0;  }
  int compute(float a) { return 0; }
}
``` 
So, when the left and right literals do not have the same types, we should check if a parameter could have a compatible type that will really keep the same test semantic.

Reply:
Make sense, and in fact, even when the issue is correct (relate to the same method taking Object for example), using parameterized tests would not result in something super clear, so let's consider only literals of the same type.